packages/embedder/JobQueueStream.ts

```diff
@@ -0,0 +1,174 @@
+import {Selectable, sql} from 'kysely'
+import ms from 'ms'
+import sleep from 'parabol-client/utils/sleep'
+import RootDataLoader from 'parabol-server/dataloader/RootDataLoader'
+import 'parabol-server/initSentry'
+import getKysely from 'parabol-server/postgres/getKysely'
+import {DB} from 'parabol-server/postgres/pg'
+import {EmbeddingsTable} from './ai_models/AbstractEmbeddingsModel'
+import {ModelManager} from './ai_models/ModelManager'
+import {createEmbeddingTextFrom} from './indexing/createEmbeddingTextFrom'
+import numberVectorToString from './indexing/numberVectorToString'
+import {updateJobState} from './indexing/updateJobState'
+
+type Job = Selectable<DB['EmbeddingsJobQueue']>
+export default class JobQueueStream implements AsyncIterableIterator<Job> {
+  private dataLoader: RootDataLoader
+  private modelManager: ModelManager
+
+  constructor(modelManager: ModelManager, dataLoader: RootDataLoader) {
+    this.modelManager = modelManager
+    this.dataLoader = dataLoader
+  }
+
+  [Symbol.asyncIterator]() {
+    return this
+  }
+  async next(): Promise<IteratorResult<Job>> {
+    const pg = getKysely()
+    let job = await pg
+      .with(
+        (cte) => cte('ids').materialized(),
+        (db) =>
+          db
+            .selectFrom('EmbeddingsJobQueue')
+            .select('id')
+            .where('state', '=', 'queued')
+            .limit(1)
+            .forUpdate()
+            .skipLocked()
+      )
+      .updateTable('EmbeddingsJobQueue')
+      .set({state: 'embedding', startAt: new Date()})
+      .where('id', '=', sql`ANY(SELECT id FROM ids)` as any)
+      .returningAll()
+      .executeTakeFirst()
+    if (!job) {
+      job = await pg
+        .with(
+          (cte) => cte('ids').materialized(),
+          (db) =>
+            db
+              .selectFrom('EmbeddingsJobQueue')
+              .select('id')
+              .where('state', '=', 'failed')
+              .where('retryAfter', '<', new Date())
+              .limit(1)
+              .forUpdate()
+              .skipLocked()
+        )
+        .updateTable('EmbeddingsJobQueue')
+        .set({state: 'embedding'})
+        .where('id', '=', sql`ANY(SELECT id FROM ids)` as any)
+        .returningAll()
+        .executeTakeFirst()
+      if (!job) {
+        console.log('JobQueueStream: no jobs found')
+        // queue is empty, so sleep for a while
+        await sleep(ms('1m'))
+        return this.next()
```

Comment:
I think about this ALL THE TIME! I wish I had an answer for you. We know it doesn't leak in prod because our gql executor uses this same pattern.
I've also tested this service for mem leaks & it works as expected. How??

If I had to take a guess, I'd say it's because `this.next()` has an arity of 0 so V8 knows there's no params it has to hold onto. We also hint to it because nothing comes after `this.next()` so my guess is when V8 figures out nothing will come after `this.next()` and it has no parameters to hold onto then it can safely GC. I've seen trampolines in the stacktrace when V8 runs out of memory, but i don't know if they use trampolines here.

It'd be really easy to test if we passed a parameter to this.next() or we put a console.log after a call to this.next()!